x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/highlight/SemanticTextHighlighter.java

```diff
@@ -0,0 +1,208 @@
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.elasticsearch.common.text.Text;
+import org.elasticsearch.common.xcontent.support.XContentMapValues;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.vectors.DenseVectorFieldMapper.DenseVectorFieldType;
+import org.elasticsearch.index.mapper.vectors.SparseVectorFieldMapper.SparseVectorFieldType;
+import org.elasticsearch.index.query.SearchExecutionContext;
+import org.elasticsearch.search.fetch.subphase.highlight.FieldHighlightContext;
+import org.elasticsearch.search.fetch.subphase.highlight.HighlightField;
+import org.elasticsearch.search.fetch.subphase.highlight.Highlighter;
+import org.elasticsearch.search.vectors.VectorData;
+import org.elasticsearch.xpack.core.ml.search.SparseVectorQueryWrapper;
+import org.elasticsearch.xpack.inference.mapper.SemanticTextField;
+import org.elasticsearch.xpack.inference.mapper.SemanticTextFieldMapper;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A {@link Highlighter} designed for the {@link SemanticTextFieldMapper}.
+ * This highlighter extracts semantic queries and evaluates them against each chunk produced by the semantic text field.
+ * It returns the top-scoring chunks as snippets, optionally sorted by their scores.
+ */
+public class SemanticTextHighlighter implements Highlighter {
+    public static final String NAME = "semantic";
+
+    private record OffsetAndScore(int offset, float score) {}
+
+    @Override
+    public boolean canHighlight(MappedFieldType fieldType) {
+        if (fieldType instanceof SemanticTextFieldMapper.SemanticTextFieldType) {
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public HighlightField highlight(FieldHighlightContext fieldContext) throws IOException {
+        SemanticTextFieldMapper.SemanticTextFieldType fieldType = (SemanticTextFieldMapper.SemanticTextFieldType) fieldContext.fieldType;
+        if (fieldType.getEmbeddingsField() == null) {
+            // nothing indexed yet
+            return null;
+        }
+
+        final List<Query> queries = switch (fieldType.getModelSettings().taskType()) {
+            case SPARSE_EMBEDDING -> extractSparseVectorQueries(
+                (SparseVectorFieldType) fieldType.getEmbeddingsField().fieldType(),
+                fieldContext.query
+            );
+            case TEXT_EMBEDDING -> extractDenseVectorQueries(
+                (DenseVectorFieldType) fieldType.getEmbeddingsField().fieldType(),
+                fieldContext.query
+            );
+            default -> throw new IllegalStateException(
+                "Wrong task type for a semantic text field, got [" + fieldType.getModelSettings().taskType().name() + "]"
+            );
+        };
```

Comment:
Does this query extraction handle when the user provides a different query for highlighting using the `highlight_query` param?

Reply:
No it does not. Although I am not sure we should support it right away. We're only looking for semantic queries so there's no point providing a different highlight query at this point.

Reply:
Agreed we shouldn't try to support this right away 👍